otwcode/otwarchive
# AO3-5702 Require language when posting a new work

# Pull Request Checklist


## Issue

https://otwarchive.atlassian.net/browse/AO3-5702

## Purpose

What does this PR do?

* Makes `language` a required field when posting (not importing) a new work.
* Adds a custom validation rather than using the Rails presence validator for two reasons:
  * Validating that the language itself exists, rather than the `language_id` being populated
  * Controlling where in the form the error message gets displayed: we wanted it at the top with any other errors, rather than crammed next to the language dropdown
* Adds specific automated tests for the new functionality
* Updates all the automated tests that were broken due to the new validation

## Testing Instructions

How can the Archive's QA team verify that this is working as you intended?

* Attempt to post a new work without choosing the language, observe that the work is not posted, and the error message appears at the top.
* Post a valid work with a language from the dropdown, and observe that it posts correctly.
* For extra credit, attempt to modify the form with a browser developer tool or similar, and replace the valid `language_id` with a made-up id. Observe that the work does not get posted and the error message appears at the top.
* Edit an existing work, deselect the language, and attempt to update the work. Observe that the work is not updated and the error message appears at the top.

## References

Are there any other relevant issues / PRs / mailing lists discussions related to this?

## Credit

*What name do you want us to use to credit your work in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?*

Enigel

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

she/her
